pkg/bpfassets/attacher/attacher.go

```diff
@@ -192,141 +159,176 @@ func attachLibbpfModule() (*bpf.Module, error) {
+		return fmt.Errorf("failed to set sample rate: %v", err)
 	}
-	err = libbpfModule.BPFLoadObject()
+	err = a.module.BPFLoadObject()
 
 	// attach to kprobe__finish_task_switch kprobe function
-	prog, err := libbpfModule.GetProgram("kepler_sched_switch_trace")
+	prog, err := a.module.GetProgram("kepler_sched_switch_trace")
 	if err != nil {
-		return libbpfModule, fmt.Errorf("failed to get kepler_sched_switch_trace: %v", err)
+		return fmt.Errorf("failed to get kepler_sched_switch_trace: %v", err)
 	} else {
 		_, err = prog.AttachGeneric()
 		if err != nil {
-			// try finish_task_switch.isra.0
 			klog.Infof("failed to attach tracepoint/sched/sched_switch: %v", err)
+		} else {
+			a.enabledSoftwareCounters[config.CPUTime] = struct{}{}
 		}
 	}
 
-	// attach softirq_entry tracepoint to kepler_irq_trace function
-	irq_prog, err := libbpfModule.GetProgram("kepler_irq_trace")
-	if err != nil {
-		klog.Warningf("could not get kepler_irq_trace: %v", err)
-		// disable IRQ metric
-		config.ExposeIRQCounterMetrics = false
-	} else {
-		_, err = irq_prog.AttachTracepoint("irq", "softirq_entry")
+	if config.ExposeIRQCounterMetrics {
+		// attach softirq_entry tracepoint to kepler_irq_trace function
+		irq_prog, err := a.module.GetProgram("kepler_irq_trace")
 		if err != nil {
-			klog.Warningf("could not attach irq/softirq_entry: %v", err)
+			klog.Warningf("could not get kepler_irq_trace: %v", err)
 			// disable IRQ metric
 			config.ExposeIRQCounterMetrics = false
+		} else {
+			_, err = irq_prog.AttachTracepoint("irq", "softirq_entry")
+			if err != nil {
+				klog.Warningf("could not attach irq/softirq_entry: %v", err)
+				// disable IRQ metric
+				config.ExposeIRQCounterMetrics = false
+			}
+			for _, event := range SoftIRQEvents {
+				a.enabledSoftwareCounters[event] = struct{}{}
+			}
 		}
 	}
 
 	// attach function
-	page_write, err := libbpfModule.GetProgram("kepler_write_page_trace")
+	page_write, err := a.module.GetProgram("kepler_write_page_trace")
 	if err != nil {
-		return libbpfModule, fmt.Errorf("failed to get kepler_write_page_trace: %v", err)
+		return fmt.Errorf("failed to get kepler_write_page_trace: %v", err)
 	} else {
 		_, err = page_write.AttachTracepoint("writeback", "writeback_dirty_folio")
 		if err != nil {
 			klog.Warningf("failed to attach tp/writeback/writeback_dirty_folio: %v. Kepler will not collect page cache write events. This will affect the DRAM power model estimation on VMs.", err)
+		} else {
+			a.enabledSoftwareCounters[config.PageCacheHit] = struct{}{}
 		}
 	}
 
 	// attach function
-	page_read, err := libbpfModule.GetProgram("kepler_read_page_trace")
+	page_read, err := a.module.GetProgram("kepler_read_page_trace")
 	if err != nil {
-		return libbpfModule, fmt.Errorf("failed to get kepler_read_page_trace: %v", err)
+		return fmt.Errorf("failed to get kepler_read_page_trace: %v", err)
 	} else {
 		if _, err = page_read.AttachGeneric(); err != nil {
```

Comment:
> Query: why some places used Attachgeneric and other places used AttachTracepoint ?

No reason. Fixed to use `AttachGeneric`.

> in case kepler_write_page_trace fails to attach, and kepler_read_page_trace succeeds there will be no entry for config.PageCacheHit in enabledSoftwareCounters

That's correct. I'm not sure that it makes sense to have only one of these 2 programs attached.
They both increase the `PageCacheHit` counter... but having only one is going to skew the results.